x-pack/legacy/plugins/spaces/server/lib/copy_to_spaces/copy_to_spaces.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract, SavedObjectsService, SavedObject } from 'src/core/server';
+import { Readable } from 'stream';
+import { SavedObjectsClientProviderOptions } from 'src/core/server';
+import { spaceIdToNamespace } from '../utils/namespace';
+import { CopyOptions, CopyResponse } from './types';
+import { getEligibleTypes } from './lib/get_eligible_types';
+import { createEmptyFailureResponse } from './lib/create_empty_failure_response';
+import { readStreamToCompletion } from './lib/read_stream_to_completion';
+
+export const COPY_TO_SPACES_SAVED_OBJECTS_CLIENT_OPTS: SavedObjectsClientProviderOptions = {
+  excludedWrappers: ['spaces'],
+};
+
+export function copySavedObjectsToSpacesFactory(
+  savedObjectsClient: SavedObjectsClientContract,
+  savedObjectsService: SavedObjectsService
+) {
+  const { importExport, types, schema } = savedObjectsService;
+  const eligibleTypes = getEligibleTypes({ types, schema });
+
+  const exportRequestedObjects = async (
+    sourceSpaceId: string,
+    options: Pick<CopyOptions, 'includeReferences' | 'objects'>
+  ) => {
+    const objectStream = await importExport.getSortedObjectsForExport({
+      namespace: spaceIdToNamespace(sourceSpaceId),
+      includeReferencesDeep: options.includeReferences,
+      objects: options.objects,
+      savedObjectsClient,
+      types: eligibleTypes,
+      exportSizeLimit: importExport.objectLimit,
+    });
+
+    return readStreamToCompletion<SavedObject>(objectStream);
+  };
+
+  const importObjectsToSpace = async (
+    spaceId: string,
+    objectsStream: Readable,
+    options: CopyOptions
+  ) => {
+    try {
+      const importResponse = await importExport.importSavedObjects({
+        namespace: spaceIdToNamespace(spaceId),
+        objectLimit: importExport.objectLimit,
+        overwrite: options.overwrite,
+        savedObjectsClient,
+        supportedTypes: eligibleTypes,
+        readStream: objectsStream,
+      });
+
+      return {
+        success: importResponse.success,
+        successCount: importResponse.successCount,
+        errors: importResponse.errors,
+      };
+    } catch (error) {
+      return createEmptyFailureResponse([error]);
+    }
+  };
+
+  const copySavedObjectsToSpaces = async (
+    sourceSpaceId: string,
+    destinationSpaceIds: string[],
+    options: CopyOptions
+  ): Promise<CopyResponse> => {
+    const response: CopyResponse = {};
+
+    const exportedSavedObjects = await exportRequestedObjects(sourceSpaceId, options);
+
+    for (const spaceId of destinationSpaceIds) {
```

Comment:
When we're copying to a single space, we could technically skip https://github.com/elastic/kibana/pull/38014/files#diff-19f7049cac00bdafdf78d0d33aacd05eR40 and just feed it directly into `importObjectsToSpace`.

Reply:
Yep, we technically could, but I felt it was cleaner and easier to read when we treat a single space the same as multiple spaces. Otherwise we end up conditionally reading the stream to completion, and it's another code path to have to reason about.

Do you think the performance difference is worth the tradeoff?

Reply:
Probably not :)